code/go/0chain.net/sharder/blockstore/fs_store.go

```diff
@@ -2,275 +2,245 @@ package blockstore
 	if err != nil {
 		return nil, err
 	}
 	return b, nil
 }
 
-// Delete - delete from the hash of the block
-func (fbs *FSBlockStore) Delete(hash string) error {
-	return common.NewError("interface_not_implemented", "FSBlockStore cannote provide this interface")
+// ReadWithBlockSummary - read the block given the block summary
+func (bStore *BlockStore) ReadWithBlockSummary(bs *block.BlockSummary) (*block.Block, error) {
+	return bStore.read(bStore, bs.Hash)
 }
 
-// DeleteBlock - delete the given block from the file system
-func (fbs *FSBlockStore) DeleteBlock(b *block.Block) error {
-	fileName := fbs.getFileName(b.Hash, b.Round)
-	err := os.Remove(fileName)
-	if err != nil {
-		return err
+// Init checks for minimum disk size, inodes requirement and assigns
+// block storer to a variable. If any error occurs during initialization
+// it will panic.
+// It will register read and write function based on whether cache config is provided or not.
+func Init(ctx context.Context, sViper *viper.Viper) {
+	logging.Logger.Info("Initializing storage")
+
+	basePath := sViper.GetString("root_dir")
+	if basePath == "" {
+		panic("root dir cannot be empty")
 	}
-	return nil
-}
 
-func (fbs *FSBlockStore) UploadToCloud(hash string, round int64) error {
-	filePath := fbs.getFileName(hash, round)
-	_, err := fbs.Minio.FPutObject(fbs.Minio.BucketName(), hash, filePath, minio.PutObjectOptions{})
+	err := hasEnoughInodesAndSize(basePath)
 	if err != nil {
-		return err
+		panic(err)
 	}
 
-	if fbs.Minio.DeleteLocal() {
-		err = os.Remove(filePath)
-		if err != nil {
-			Logger.Error("Failed to delete block which is moved to cloud", zap.Int64("round", round), zap.String("path", filePath))
-		}
-		Logger.Info("Local block successfully deleted, moved to cloud", zap.Int64("round", round), zap.String("path", filePath))
+	bStore := new(BlockStore)
+	bStore.blockMetadataProvider = datastore.GetEntityMetadata("block")
+	bStore.basePath = basePath
+
+	cViper := sViper.Sub("cache")
+	if cViper != nil {
+		bStore.cache = initCache(cViper)
 	}
-	return nil
-}
 
-func (fbs *FSBlockStore) DownloadFromCloud(hash string, round int64) error {
-	filePath := fbs.getFileName(hash, round)
-	return fbs.Minio.FGetObject(fbs.Minio.BucketName(), hash, filePath, minio.GetObjectOptions{})
-}
+	switch {
+	default:
+		bStore.write = func(bStore *BlockStore, hash string, b *block.Block) error {
+			return bStore.writeToDisk(hash, b)
+		}
+
+		bStore.read = func(bStore *BlockStore, hash string) (*block.Block, error) {
+			return bStore.readFromDisk(hash)
+		}
+	case cViper != nil:
+		bStore.write = func(bStore *BlockStore, hash string, b *block.Block) error {
+			err := bStore.writeToDisk(hash, b)
+			if err != nil {
+				return err
+			}
+
+			go func() {
+				if err := bStore.writeBlockToCache(hash, b); err != nil {
+					logging.Logger.Error(err.Error())
+				}
+			}()
+
+			return nil
+		}
+
+		bStore.read = func(bStore *BlockStore, hash string) (*block.Block, error) {
```

Comment:
the same use no-op cache in case you don't need it, then you can define `read` in struct as a func